x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization2/CategorizationPartOfSpeechDictionary.java

```diff
@@ -0,0 +1,158 @@
+                continue;
+            }
+            String[] split = line.split(PART_OF_SPEECH_SEPARATOR);
+            if (split.length != 2) {
+                throw new IllegalArgumentException(
+                    "Unexpected format in line [" + line + "]: expected one [" + PART_OF_SPEECH_SEPARATOR + "] separator"
+                );
+            }
+            if (split[0].isEmpty()) {
+                throw new IllegalArgumentException(
+                    "Unexpected format in line [" + line + "]: nothing preceding [" + PART_OF_SPEECH_SEPARATOR + "] separator"
+                );
+            }
+            if (split[1].isEmpty()) {
+                throw new IllegalArgumentException(
+                    "Unexpected format in line [" + line + "]: nothing following [" + PART_OF_SPEECH_SEPARATOR + "] separator"
+                );
+            }
+            String lowerCaseWord = split[0].toLowerCase(Locale.ROOT);
+            partOfSpeechDictionary.put(lowerCaseWord, PartOfSpeech.fromCode(split[1].charAt(0)));
+            maxLength = Math.max(maxLength, lowerCaseWord.length());
+        }
+        maxDictionaryWordLength = maxLength;
+    }
+
+    // TODO: now we have this in Java, perform this operation in Java for anomaly detection categorization instead of in C++.
+    // (It could maybe be incorporated into the categorization analyzer and then shared between aggregation and anomaly detection.)
+    /**
+     * Find the part of speech (noun, verb, adjective, etc.) for a supplied word.
+     * @return Which part of speech does the supplied word represent? {@link PartOfSpeech#NOT_IN_DICTIONARY} is returned
+     *         for words that aren't in the dictionary at all.
+     */
+    public PartOfSpeech getPartOfSpeech(String word) {
+        if (word.length() > maxDictionaryWordLength) {
+            return PartOfSpeech.NOT_IN_DICTIONARY;
+        }
+        // This is quite slow as it creates a new string for every lookup. However, experiments show
+        // that trying to do case-insensitive comparisons instead of creating a lower case string is
+        // even slower.
+        return partOfSpeechDictionary.getOrDefault(word.toLowerCase(Locale.ROOT), PartOfSpeech.NOT_IN_DICTIONARY);
+    }
+
+    /**
+     * @return Is the supplied word in the dictionary?
+     */
+    public boolean isInDictionary(String word) {
+        return getPartOfSpeech(word) != PartOfSpeech.NOT_IN_DICTIONARY;
+    }
+
+    public static CategorizationPartOfSpeechDictionary getInstance() throws IOException {
+        if (instance != null) {
+            return instance;
+        }
+        synchronized (INIT_LOCK) {
+            if (instance == null) {
+                try (InputStream is = CategorizationPartOfSpeechDictionary.class.getResourceAsStream(DICTIONARY_FILE_PATH)) {
+                    instance = new CategorizationPartOfSpeechDictionary(is);
+                }
+            }
+            return instance;
+        }
+    }
+}
```

Comment:
I think it's best not to add it to the same circuit breaker used by the rest of the aggregation.

Although it's large it's effectively static data, so it would make most sense to include it with the ["Accounting requests circuit breaker"](https://www.elastic.co/guide/en/elasticsearch/reference/current/circuit-breaker.html#accounting-circuit-breaker) rather than the ["Request circuit breaker"](https://www.elastic.co/guide/en/elasticsearch/reference/current/circuit-breaker.html#request-circuit-breaker).  But if `indices.breaker.total.use_real_memory` is set to `true`, which it is by default, then that "memory usage of things held in memory that are not released when a request is completed" will take it into account automatically.

I guess we _could_ try to explicitly add it into the "Accounting requests circuit breaker" for the case where real memory circuit breaking is disabled.  But this will be messy within the code as the code is written on the basis that what the docs refer to as "memory usage of things held in memory that are not released when a request is completed" is actually field data related to Lucene indices.

The [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/circuit-breaker.html#parent-circuit-breaker) also say about the total memory all circuit breakers can use: "Defaults to 70% of JVM heap if `indices.breaker.total.use_real_memory` is false. If `indices.breaker.total.use_real_memory` is true, defaults to 95% of the JVM heap."  So that implies that if you don't use the real memory circuit breaker to measure fixed overheads then you have to allow some space for unmeasured fixed overheads.  So I think this dictionary can be treated as one of those fixed overheads that either gets captured by the real memory circuit breaker or by implicitly reserving a percentage of memory.